src/Storages/MergeTree/IMergeTreeDataPart.cpp

```diff
@@ -424,53 +424,60 @@ void IMergeTreeDataPart::setColumns(const NamesAndTypesList & new_columns)
 
 void IMergeTreeDataPart::removeIfNeeded()
 {
-    if (state == State::DeleteOnDestroy || is_temp)
-    {
-        try
-        {
-            auto path = getFullRelativePath();
+    if (!is_temp && state != State::DeleteOnDestroy)
+        return;
 
-            if (!volume->getDisk()->exists(path))
-                return;
+    try
+    {
+        auto path = getFullRelativePath();
 
-            if (is_temp)
-            {
-                String file_name = fileName(relative_path);
+        if (!volume->getDisk()->exists(path))
+            return;
 
-                if (file_name.empty())
-                    throw Exception("relative_path " + relative_path + " of part " + name + " is invalid or not set", ErrorCodes::LOGICAL_ERROR);
+        if (is_temp)
+        {
+            String file_name = fileName(relative_path);
 
-                if (!startsWith(file_name, "tmp") && !endsWith(file_name, ".tmp_proj"))
-                {
-                    LOG_ERROR(
-                        storage.log,
-                        "~DataPart() should remove part {} but its name doesn't start with \"tmp\" or end with \".tmp_proj\". Too "
-                        "suspicious, keeping the part.",
-                        path);
-                    return;
-                }
-            }
+            if (file_name.empty())
+                throw Exception(ErrorCodes::LOGICAL_ERROR, "relative_path {} of part {} is invalid or not set", relative_path, name);
 
-            if (parent_part)
+            if (!startsWith(file_name, "tmp") && !endsWith(file_name, ".tmp_proj"))
             {
-                std::optional<bool> keep_shared_data = keepSharedDataInDecoupledStorage();
-                if (!keep_shared_data.has_value())
-                    return;
-                projectionRemove(parent_part->getFullRelativePath(), *keep_shared_data);
+                LOG_ERROR(
+                    storage.log,
+                    "~DataPart() should remove part {} but its name doesn't start with \"tmp\" or end with \".tmp_proj\". Too "
+                    "suspicious, keeping the part.",
+                    path);
+                return;
             }
-            else
-                remove();
+        }
 
-            if (state == State::DeleteOnDestroy)
-            {
-                LOG_TRACE(storage.log, "Removed part from old location {}", path);
-            }
+        if (parent_part)
+        {
+            std::optional<bool> keep_shared_data = keepSharedDataInDecoupledStorage();
+            if (!keep_shared_data.has_value())
+                return;
+            projectionRemove(parent_part->getFullRelativePath(), *keep_shared_data);
         }
-        catch (...)
+        else
+            remove();
+
+        if (state == State::DeleteOnDestroy)
         {
-            tryLogCurrentException(__PRETTY_FUNCTION__);
+            LOG_TRACE(storage.log, "Removed part from old location {}", path);
         }
     }
+    catch (...)
```

Comment:
Probably for remote disks we have to add smart retries in future (some exception codes can be retried).